ironic/drivers/modules/drac/management.py

```diff
@@ -302,14 +315,182 @@ def set_boot_device(node, device, persistent=False):
+        else:
+            LOG.error("Failed to clear job queue, node : %(node)s " %
+                     {'node': task.node.uuid})
+        return delete_job_response
+
+    @METRICS.timer('DracRedfishManagement.reset_idrac')
+    @base.clean_step(priority=0)
+    def reset_idrac(self, task):
+        """Reset the iDRAC.
+
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: DracOperationError on an error from python-dracclient.
+        """
+        system = redfish_utils.get_system(task.node)
+        for manager in system.managers:
+            try:
+                oem_manager = manager.get_oem_extension('Dell')
+            except sushy.exceptions.OEMExtensionNotFoundError as e:
+                error_msg = (_("Search for Sushy OEM extension package "
+                           "'sushy-oem-idrac' failed for node %(node)s. "
+                           "Ensure it is installed. Error: %(error)s") %
+                         {'node': task.node.uuid, 'error': e})
+                LOG.error(error_msg)
+                raise exception.RedfishError(error=error_msg)
+            try:
+                reset_job_response = oem_manager.reset_idrac(
+                                                manager=manager)
+            except sushy.exceptions.SushyError as e:
+                LOG.debug("Sushy OEM extension Python package "
+                          "sushy-oem-idrac failed to reset idrac"
+                          "for %(node)s, Error : %(error)s" %
+                          {'node':task.node.uuid, 'error':e})
+        url = oem_manager._conn._url
+        redfish_node_ip = re.search(r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}',
+                                    url).group()
+        if reset_job_response.status_code == _RESET_JOB_RESPONSE_CODE:
+            LOG.info("idrac reset success for node %(node)s via OEM" %
+                    {'node': task.node.uuid})
+            LOG.debug("iDRAC was reset, waiting for return to operational state")
+            redfish_utils.wait_for_host(redfish_node_ip)
+            LOG.info("The iDRAC has become pingable")
+            LOG.info("Waiting for the iDRAC to become ready")
+            redfish_utils.wait_until_idrac_is_ready(oem_manager,
+                        retries=IDRAC_IS_READY_RETRIES,
+                        retry_delay=IDRAC_IS_READY_RETRY_DELAY_SEC)
+        else:
+            LOG.error("Failed to reset idrac for %(node)s" %
+                    {'node': task.node.uuid})
+        return reset_job_response
+
+    @METRICS.timer('DracRedfishManagement.known_good_state')
+    @base.clean_step(priority=0)
+    def known_good_state(self, task):
+        """Reset the iDRAC, Clear the job queue.
+
+        :param task: a TaskManager instance containing the node to act on.
+        :returns: None if it is completed.
+        :raises: DracOperationError on an error from python-dracclient.
+        """
```

Comment:
I think almost all of the code in this method could be remove and replaced with 2 lines that call into reset_idrac and clear_job_queue.  We should call reset_idrac first since that may get it into a state where clear_job_queue succeeds where otherwise it would have failed.